### PR TITLE
Settings export/import

### LINKED ISSUE #3176: Import/Export app's settings
Screenshot_2022-02-18-14-54-59

source: Amaze FM's playstore review

### PR DESCRIPTION
**Pull request by VishnuSanal**
#### Issue tracker   
Fixes #3176

#### Automatic tests
- [ ] Added test cases
  
#### Manual tests
- [x] Done  

- Device: Pixel 3A Emulator
- OS: Android Tiramisu x86_64

#### Build tasks success
Successfully running following tasks on local:
- [x] `./gradlew assembledebug`
- [x] `./gradlew spotlessCheck`

### CODE REVIEW

File: app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+                val map: Map<String?, Any> = Gson().fromJson(
+                    stringBuilder.toString(),
+                    type
+                )
+
+                val editor: SharedPreferences.Editor? =
+                    PreferenceManager.getDefaultSharedPreferences(getActivity()).edit()
+
+                for ((key, value) in map) try {
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt:149]: Use when syntax

File: app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+
+        if (requestCode == IMPORT_BACKUP_FILE &&
+            resultCode == Activity.RESULT_OK &&
+            data != null && data.data != null
+        ) {
+            val uri = data.data
+
+            Log.e(TAG, "read import file: $uri")
+
+            try {
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt:127]: This has a few try catch nested, you should collect all errors and provide the user with error based feedback. Maybe print the exception message to a toast?

File: app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+
+        if (requestCode == IMPORT_BACKUP_FILE &&
+            resultCode == Activity.RESULT_OK &&
+            data != null && data.data != null
+        ) {
+            val uri = data.data
+
+            Log.e(TAG, "read import file: $uri")
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt:125]: Why the log? why is this an error log?

File: app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+
+        try {
+
+            val file = File(context?.cacheDir?.absolutePath + File.separator + "amaze_backup.json")
+
+            val fileWriter = FileWriter(file)
+
+            fileWriter.append(gsonString)
+
+            Log.e(TAG, "wrote export to: ${file.absolutePath}")
```
Comment (EmmanuelMess) [inline app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt:62]: Logs error.

**Review by EmmanuelMess — CHANGES_REQUESTED** [review]

**Comment by EmmanuelMess** [comment]
Please add a test that exports from one version and imports to another. How does this manage adding settings?

File: app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt (hunk 3 above)
Comment (VishnuSanal) [reply app/src/main/java/com/amaze/filemanager/ui/fragments/preferencefragments/BackupPrefsFragment.kt:125]: whoops, my bad, it should be a `Log#i`

**Comment by VishnuSanal** [comment]
> Please add a test that exports from one version and imports to another.

sure.

> How does this manage adding settings?

settings would be imported only if it is present in the backup file, else it is left to the user to set it manually.

**Comment by VishalNehra** [comment]
@VishnuSanal  can you fix the codacy failures here please?

**Comment by VishnuSanal** [comment]
> @VishnuSanal can you fix the codacy failures here please?

sure, but can we ignore the missing documentation warnings?

**Comment by VishnuSanal** [comment]
> > @VishnuSanal can you fix the codacy failures here please?
> 
> sure, but can we ignore the missing documentation warnings?

NVM, fixed them.

**Review by VishalNehra — APPROVED** [review]

**Comment by bananaking24** [comment]
how do i use it im trying to install amaze but idk howwww

**Pull request merged**